.github/workflows/build_test.yml

```diff
@@ -24,122 +24,164 @@ jobs:
-      python-version: ["3.8"]
-      cloud-provider: [aws, azure, gcp]
-   steps:
-    - uses: actions/checkout@v2
-    - name: Set up Python
-      uses: actions/setup-python@v2
-      with:
-        python-version: ${{ matrix.python-version }}
-    - name: Display Python version
-      run: python -c "import sys; print(sys.version)"
-    - name: Setup parameters file
-      shell: bash
-      env:
-        PARAMETERS_SECRET: ${{ secrets.PARAMETERS_SECRET }}
-      run: |
-        gpg --quiet --batch --yes --decrypt --passphrase="$PARAMETERS_SECRET" \
-        .github/workflows/parameters/parameters_${{ matrix.cloud-provider }}.py.gpg > tests/parameters.py
-    - name: Upgrade setuptools, pip and wheel
-      run: python -m pip install -U setuptools pip wheel
-    - name: Install tox
-      run: python -m pip install tox
-    - name: List installed packages
-      run: python -m pip freeze
-    - name: Run tests
-      run: python -m tox -e "py${PYTHON_VERSION/\./}" --skip-missing-interpreters false
-      env:
-        PYTHON_VERSION: ${{ matrix.python-version }}
-        PYTEST_ADDOPTS: -vvv --color=yes --tb=short
-    - name: Combine coverages
-      run: python -m tox -e coverage --skip-missing-interpreters false
-      shell: bash
-    - uses: actions/upload-artifact@v2
-      with:
-        name: coverage_${{ matrix.os.download_name }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
-        path: |
-          .tox/.coverage
-          .tox/coverage.xml
+  test-dialect:
+    name: Test dialect ${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
+    needs: lint
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [
+          ubuntu-latest,
+          macos-latest,
+          windows-latest,
+          ]
+        python-version: ["3.8"]
+        cloud-provider: [
+          aws,
+          azure,
+          gcp,
+          ]
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Upgrade pip and prepare environment
+        run: |
+          python -m pip install -U pip
+          python -m pip install -U hatch
+          python -m hatch env create default
+      - name: Setup parameters file
+        shell: bash
+        env:
+          PARAMETERS_SECRET: ${{ secrets.PARAMETERS_SECRET }}
+        run: |
+          gpg --quiet --batch --yes --decrypt --passphrase="$PARAMETERS_SECRET" \
+          .github/workflows/parameters/parameters_${{ matrix.cloud-provider }}.py.gpg > tests/parameters.py
+      - name: Run tests
+        run: hatch run test-dialect
+      - uses: actions/upload-artifact@v4
+        with:
+          name: coverage.xml_dialect-${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
+          path: |
+            ./coverage.xml
```

Comment:
upload is not needed if we don't have coverage

Reply:
You're right.
I wonder id we should have raport to look at for future PR's